New: code path analysis.

mysticatea opened the pull request:
Fixes #3530 

- [x] Decide API (new events style v.s. `context.getCodePath()` style)
- [x] Add JSDoc comments.
- [x] Add more tests.
  - [x] logical expressions.
  - [x] if statements.
  - [x] switch statements.
  - [x] try-catch-finally statements.
  - [x] while statements.
  - [x] do-while statements.
  - [x] for statements.
  - [x] for-in/of statements.
  - [x] functions.
  - [x] default parameters.
  - [x] default values of destructuring assignments
- [x] Add documents.
- [x] Remake some rules with the code path analysis.
  - [x] `no-unreachable`
  - [x] `no-fallthrough`
  - [x] `constructor-super`
  - [x] `no-this-before-super`
  - [x] `consistent-return`

mysticatea commented:
Currently, I'm adding tests and fixing comments around `switch`...

nzakas commented on lib/code-path-analysis/code-path-analyzer.js:
```diff
@@ -0,0 +1,436 @@
+"use strict";
```
Be sure to update this file with the header comments like other files have.

nzakas replied:
See: http://eslint.org/docs/developer-guide/code-conventions.html

mysticatea replied:
Yeah, I have not written most comments yet...
I will add that.

nzakas commented on lib/code-path-analysis/code-path-analyzer.js:
```diff
@@ -0,0 +1,436 @@
+var CodePath = require("./code-path");
+var CodePathSeg = require("./code-path-seg");
+
+/**
```
Maybe separate debugging helpers into a separate file?

mysticatea commented:
I'm adding tests for `try-catch-finally` currently.

I'm considering paths in `try` blocks.

```js
class A {
    constructor(a) {
        try {
            super(a);
        } catch (err) {
        }

        this.foo();
    }
}
```

This code does not throw any error before calling `super(a);`.
My plan is creating two paths from `try` block to `catch` block.

- The first throwable node (e.g. `Identifier` as a reference)
- The last segment in `try` block.
- (and at `throw` statements)

Because those are at opposite poles. I guess most rules is good enough to use the two paths.
But in this case, `a` is `Identifier` as a reference, but not throwable. 
In order to detect whether the Identifier is throwable or not, resolving names is needed.

nzakas commented:
As far as I know, you can't use `this` if `super()` throws.

Regarding whether to fork or not, is there any existing code path analysis tool that might be useful to look at? For instance, maybe Flow does this and we could figure out the right way to handle it based on that?

@jeffmo are you all doing this type of code path analysis? Any insights?

jeffmo commented:
I'm pretty sure we don't track partial initialization state right now except in a few very trivial circumstances. It's a TODO for us, but probably not in the short term.

This is pretty neat, though. I imagine when we do get to this we'll model it like any other variable initialization and ask the user to consider "potentially uninitialized" for `this` as well. `super()` isn't a normal "call" by pretty much any definition...so I wouldn't feel too bad about special-casing it as 'throwable'.

P.S. Just tossing in the spec text confirming what you said regarding `this` not being initialized if `super()` throws ("SuperCall : super Arguments" - Steps 8-10):

http://www.ecma-international.org/ecma-262/6.0/index.html#sec-super-keyword-runtime-semantics-evaluation

mysticatea commented:
@nzakas @jeffmo Thank you very much.

escope have completed scope analysis in this time, so I guessed we can detect whether the Identifier is throwable or not. But escope does not have `O(1)` getter for variables...

So I will do not detecting that to start. I will think below as throwable:

- All `Identifier` as a reference.
- All `MemberExpression` (can throw a `null` access error, or might be a getter)
- All `CallExpression` (includes `super()`)
- All `NewExpression`

nzakas commented:
Sounds good.

nzakas commented on lib/eslint.js:
```diff
@@ -807,20 +811,18 @@ module.exports = (function() {
                     emitCommentsEnter(comments.leading);
                     node.parent = parent;
-                    api.emit(node.type, node);
+                    codePathAnalyzer.enterNode(node);
```
I'm not sure it's a good idea to have `codePathAnalyzer` emit the normal node events. Maybe have `enterNode()` return the code path and segment so eslint can emit them?

mysticatea replied:
OK. There are some note:

- Code path events are fired only when it's changed from the previous one.
- An ast node has several code path segments each AST node in `finally` blocks. So it's possible to fire several `onCodePathSegment` events on a AST node.
- There are both events which are fired before and after the normal AST event.

So I guess the return value is such as:

```js
{
    "before": [
        {"type": "onCodePath:exit", "args": [codePath1]},
        {"type": "onCodePath", "args": [codePath2]},
        {"type": "onCodePathSegment", "args": [segment1, codePath2]},
        {"type": "onCodePathSegment", "args": [segment2, codePath2]}
        // ...
    ],
    "after": [] // similar to `before`
}
```

nzakas replied:
Hmm, that's not very pretty. :)

I'm still concerned that we different parts emitting for different things. Can the code path analyzer just emit events for code paths and leave emitting node events in eslint?

mysticatea replied:
I have three ideas.

1. To separate methods.

  ```js
  codePathAnalyzer.enterNode(node);
  api.emit(node.type, node);
  codePathAnalyzer.leaveEnteringNode(node);

  //....

  codePathAnalyzer.enterLeavingNode(node);
  api.emit(node.type + ":exit", node);
  codePathAnalyzer.leaveNode(node);
  ```

2. A callback.

  ```js
  codePathAnalyzer.enterNode(node, function() {
      api.emit(node.type, node);
  });

  //....

  codePathAnalyzer.leaveNode(node, function() {
      api.emit(node.type + ":exit", node);
  });
  ```

3. The decorator pattern. 

![image](https://cloud.githubusercontent.com/assets/1937871/9675875/652c78e0-52fe-11e5-9cf6-0c72a2ea8bc0.png)

  ```js
  var visitor = new CommentTraverser(new CodePathAnalyzer(new NodeTraverser(api), api), api);

  //....

  visitor.enterNode(node);

  //....

  visitor.leaveNode(node);
  ```

nzakas replied:
Ooh, option #3 is a really nice and clean way of doing it. It would also make further changes a lot easier.

ilyavolodin replied:
Actually I think option 3 will give us a very nice doorway for selectors too. We would just need to add SelectorAnalyzer (or whatever we decide to call that) at the same level as `CommentTraverser` and `CodePathAnalyzer`.

nzakas commented on lib/code-path-analysis/code-path-analyzer.js:
```diff
@@ -0,0 +1,350 @@
+
+        // Emits node events.
+        debug.dumpState(node.type, state);
+        this.emitter.emit(
```
Not entirely sure about padding the segment and codepath in as part of the node handler. I'm worried we might want to pass in more than just codepath information at some point, so maybe we should have some sort of node meta object that contains codepath info? That way we can always add more properties if we want to pass in more info. What do you think?

mysticatea replied:
Oh, Yes my bad.

I will add `context.getCodePath()` similar to `context.getScope()`, instead.

nzakas commented:
@mysticatea still working on this?

mysticatea commented:
Yes, I'm working on this. I'm sorry for slowly.
I have time in weekends mainly (I have late on my job now) and I'm setting higher priority to fix bugs.

nzakas commented:
@mysticatea do you want some help to finish this?

mysticatea commented:
Thank you.
I completed implementation and I'm writing documents now.

mysticatea commented:
Done, ready for review!

I have run the rewritten rules in some repositories (eslint, escope, espree, npm, and mine), then I didn't find wrong results.

This PR contains:

- Code Path Analyzer's code (`lib/code-path-analysis/*`).
- Code Path Analyzer's tests (`tests/lib/code-path-analysis/*` and `tests/fixtures/code-path-analysis/*`).
- 5 rules are rewritten with code paths.
- Some code modifications that rewritten rules have found.
- Polyfill for `Function.prototype.bind` of PhantomJS.
- A document (`docs/developer-guide/working-with-rules.md`)

----

Main files:

- `code-path.js`
- `code-path-analyzer.js`
- `code-path-segment.js`
- `code-path-state.js`
- `debug-helpers.js`
- `fork-context.js`
- `id-generator.js`

Most process is in `code-path-analyzer.js` and `code-path-state.js`.
`code-path-analyzer.js` traverses AST and gives information to `code-path-state.js`.
`code-path-state.js` manages the state of a code path with given information from `code-path-analyzer.js`.
`code-path.js` and `code-path-segment.js` are public interface. 
`debug-helpers.js`, `fork-context.js`, and `id-generator.js` are internal helpers.

gyandeeps commented:
Before we can start reviewing the code, can make sure your branch is rebased with the latest master.

gyandeeps commented on lib/eslint.js:
```diff
@@ -461,6 +462,7 @@ function getRuleReplacementMessage(ruleId) {
         var newRules = replacements.rules[ruleId];
         return "Rule \'" + ruleId + "\' was removed and replaced by: " + newRules.join(", ");
     }
+    return null;
```
Just curious, why was this change made?

mysticatea replied:
Because new `consistent-return` rule had been warning this function: `Expected to return a value at the end of this function.`

gyandeeps commented:
Quick question:
* What happens if not all code paths return? Like the example below
```js
function test(x){
  if(x === 1){
    console.log("do nothing");
  }
  else{
    return true;
  }
}
```

* Similar question but for switch cases where if I dont define a default case?
* Can you also share the perf run for before and after this code?

Side note: To be completely honest here (:smile:) I dont think I can completely digest this whole logic in one go. I have to run through and play around with code to grasp how all this works.

ilyavolodin commented on lib/code-path-analysis/code-path-analyzer.js:
```diff
@@ -0,0 +1,603 @@
+ * @returns {boolean} `true` if the node is a `case` node.
+ */
+function isCaseNode(node) {
+    return Boolean(node.test);
```
Would this also return true for `if` statements?

nzakas replied:
Yeah, better to use `node.type === "SwitchCase"`

mysticatea replied:
I want to write here: `node.test != null` :(
This function's intention is `this node is not "default:".`

Also, I don't like:

```js
if (...) {
    return true;
} else {
    return false;
}
```

ilyavolodin commented:
I completely agree with @gyandeeps this is just too much code to go through in a single code review. @mysticatea any ideas if we can somehow split this up into multiple PRs? For example we should be able to easily separate rule updates into it's own PR. Anything else that can be pulled out?
This is one of the most complicated features in the whole code-base. I think it deserves thorough code review, and I'm very excited about possibilities that this opens up. But going through 130 files is just too much.

ilyavolodin commented on lib/code-path-analysis/code-path-analyzer.js:
```diff
@@ -0,0 +1,603 @@
+/**
+ * Gets a label text from a given loop node.
+ *
+ * @param {ASTNode} node - A loop node to get.
```
Could you specify which nodes are acceptable? I assume `for`, `while` and `do`?

nzakas commented on docs/developer-guide/working-with-rules.md:
````diff
@@ -311,6 +312,117 @@ Keep in mind that comments are technically not a part of the AST and are only at
+
+> **Note:** That code path has not completed. If you want to get the completed code path, you can use events which will be described later.
+
+```ts
````
I don't think `ts` will work when we generate the site.

nzakas commented on docs/developer-guide/working-with-rules.md:
```diff
@@ -311,6 +312,117 @@ Keep in mind that comments are technically not a part of the AST and are only at
+         * @param {ASTNode} node - The current node.
+         * @returns {void}
+         */
+        "onCodePath": function(codePath, node) {
```
I wonder if we should change the naming of these events to:

* `onCodePathStart`
* `onCodePathEnd`
* `onCodePathSegmentStart`
* `onCodePathSegmentEnd`

That will make it clearer that these are not AST nodes.

nzakas commented on docs/developer-guide/working-with-rules.md:
```diff
@@ -311,6 +312,117 @@ Keep in mind that comments are technically not a part of the AST and are only at
+ESLint is analyzing code paths while traversing AST. You can access that code path objects with `context.getCodePath()`.
+`context.getCodePath()` returns a `CodePath` object of the current location.
+
+> **Note:** That code path has not completed. If you want to get the completed code path, you can use events which will be described later.
```
This seems like it could potentially cause a lot of confusion. If the `CodePath` object isn't completed when someone calls `getCodePath()`, that could lead to errors, correct?

If yes, is there a way to do the analysis first? I'm a little worried that we could see a lot of errors relating to `getCodePath()` if people expect the object to be complete (like `getScope()`).

mysticatea replied:
No, it does not raise any error.
Just on the way. e.g. "the code path object does not have `finalSegs` content."

nzakas commented on docs/developer-guide/working-with-rules.md:
````diff
@@ -311,6 +312,117 @@ Keep in mind that comments are technically not a part of the AST and are only at
+    reachable: boolean;
+}
+```
+
````
Can you explain the difference between a `CodePath` and a `CodePathSegment`?

mysticatea replied:
These are similar to `LinkedList<T>` and `LinkedListNode<T>`.
`CodePath` is references to the initial and the final of a code path.
`CodePathSegment` is parts of a code path.

nzakas commented on docs/developer-guide/working-with-rules.md:
```diff
@@ -311,6 +312,117 @@ Keep in mind that comments are technically not a part of the AST and are only at
+
+#### Events for Code Paths
+
+There are 5 events for Code Paths. These events are written as similar to callbacks of AST.
```
Wording cleanup:

"There are five events related to code paths, and you can define event handlers in rules."

nzakas commented on lib/code-path-analysis/code-path-analyzer.js:
```diff
@@ -0,0 +1,603 @@
+}
+
+/**
+ * Updates the code path due to the position of a given node in the parent node
```
Can you explain more? It's not clear what this method does.

nzakas commented on lib/code-path-analysis/code-path-segment.js:
```diff
@@ -0,0 +1,218 @@
+     * An array of the next segments.
+     * @type {CodePathSegment[]}
+     */
+    this.nextSegs = [];
```
I think we should use "segments" everywhere to be consistent:

```
nextSegs -> nextSegments
prevSegs -> previousSegments
allNextSegs -> allNextSegments
allPrevSegs -> allPreviousSegments
```

nzakas commented on lib/code-path-analysis/code-path-state.js:
```diff
@@ -0,0 +1,1111 @@
+ * @param {any[]} xs - An array to get.
+ * @returns {any|null} The last element, or `null` if the array is empty.
+ */
+function lastOf(xs) {
```
This is in a few files - may be best to extract into a utility?

nzakas commented on lib/formatters/checkstyle.js:
```diff
@@ -41,7 +41,8 @@ function xmlEscape(s) {
             case "'":
                 return "&apos;";
-            // no default
+            default:
```
Why this change?

mysticatea replied:
Because new `consistent-return` rule had been warning this function: `Expected to return a value at the end of this function.`

nzakas commented:
This is an amazing amount of work, thanks @mysticatea! 

I agree with @ilyavolodin, I think we should split this into a some smaller pull requests to make it easier for people to review.

I'm also a little bit concerned about the note regarding `getCodePath()` returning an incomplete `CodePath` object, as that seems like it could be confusing to people and potentially cause errors.

We're also going to have to do a lot of documentation around how this works. I'm not sure the difference between a `CodePath` and `CodePathSegment`, is there any place I can go to read more about this approach?

mysticatea commented:
@gyandeeps 

> What happens if not all code paths return? Like the example below

If all paths are not returned, it addresses that the final segment is returned.

![image](https://cloud.githubusercontent.com/assets/1937871/10441589/b47b583a-7185-11e5-883b-8b152576d5d7.png)

> Similar question but for switch cases where if I dont define a default case?

There are test cases here: `tests\fixtures\code-path-analysis\switch--cases-*`
It goes the next code after walking all case tests.

> Can you also share the perf run for before and after this code?

BEFORE:

```
CPU Speed is 3392 with multiplier 7500000
Performance Run #1:  1972.3731480000001ms
Performance Run #2:  1957.460232ms
Performance Run #3:  1983.6695949999998ms
Performance Run #4:  1969.877751ms
Performance Run #5:  1982.227052ms
Performance budget ok:  1972.3731480000001ms (limit: 2211.0849056603774ms)


Load performance Run #1:  234.612575ms
Load performance Run #2:  233.000996ms
Load performance Run #3:  232.582933ms
Load performance Run #4:  230.61668ms
Load performance Run #5:  232.313381ms

Load Performance median :  232.582933ms
```

AFTER:

```
CPU Speed is 3392 with multiplier 7500000
Performance Run #1:  2168.29696ms
Performance Run #2:  2165.601437ms
Performance Run #3:  2187.562861ms
Performance Run #4:  2165.585137ms
Performance Run #5:  2159.003894ms
Performance budget ok:  2165.601437ms (limit: 2211.0849056603774ms)


Load performance Run #1:  242.883576ms
Load performance Run #2:  239.578617ms
Load performance Run #3:  238.709589ms
Load performance Run #4:  240.879894ms
Load performance Run #5:  238.128226ms

Load Performance median :  239.578617ms
```

About 10% slower :cry: 

----

There are many test cases in `tests\fixtures\code-path-analysis`.
Each test case has js code and DOT code to visualize.
If we copy and paste DOT code to http://www.webgraphviz.com/ , we can look at a code path as dygraph.

Also, debug mode (`DEBUG="eslint:code-path"`) outputs DOT codes of each code path.

mysticatea commented:
@ilyavolodin @nzakas 

The main purpose of 5 rules was that I confirm the ability of this feature.
So I can separate 5 rules from this PR.
But I thought also those are good examples which help to understand this feature.
I think a better way is that I separate this commit for analysis and each rule.
(93/130 files are just test case!)

Thought?

> I'm also a little bit concerned about the note regarding getCodePath() returning an incomplete CodePath object, as that seems like it could be confusing to people and potentially cause errors.

`getCodePath()` returns the result of until the current position.
It's incomplete, but I think it's enough for rules.
If a rule needs complete code path, it can use "Program:exit" like some rules for variables. Or it just can use "onCodePathEnd" handler.

I can rewrite that it creates code paths in preprocessing.
In that case, also mapping between nodes and code path segments is needed to save additional information of code paths for each rule. A node needs to map to different segment in entering and exiting. And a node can map to several segments.

> We're also going to have to do a lot of documentation around how this works.

I agree.
I want to make examples page.
There are many pairs of js code and DOT code in `tests\fixtures\code-path-analysis`, so I think maybe I can auto-generate examples.

Thought?

-----

I will work on improving that item that was pointed out today (JST).

mysticatea commented:
I'm sorry, I failed to get time today.

nzakas commented:
> getCodePath() returns the result of until the current position.
It's incomplete, but I think it's enough for rules.
If a rule needs complete code path, it can use "Program:exit" like some rules for variables. Or it just can use "onCodePathEnd" handler.

I'm just worried this will be confusing for people, because nothing else is incomplete when accessed via `context`. I can see two possible ways to fix this:

1. Just don't have `getCodePath()` (can we still make the correct changes to the rules if we remove this?)
1. Create the code paths *before* we start the traversal to apply rules (the same way `escope` works)

Do either of those options work?

We're also adding quite a bit (200ms) in performance overhead - on a large codebase, that could make the process a lot slower. I'd like to see if we can do some performance optimization before releasing.

mysticatea commented:
OK, I favor 1.
Rules can work by managing the current segments with `onCodePathSegmentStart` and `onCodePathSegmentEnd`.
That 2nd needs mapping from all nodes to respective segments is troublesome...

mysticatea commented:
I tried to separate a commit to the analysis and rules.
I will squash commits after review, at the time, I will rethink to separate rules.

Rules show usage of this feature.... but, I'm working on removing `context.getCodePath()`, so it will be changed a bit.

mysticatea commented:
I removed `context.getCodePath()`.

mysticatea commented:
Though I'm trying to write `no-null-dereference` rule with this feature now, then I found wrong path on a combo of LogicalExpression and IfStatement (statements which has `test` property).
For example, In `&&` case, the path of `LogicalExpression.left` should not go to the `IfStatement.consequent` chunk. But the path is going there in the current implementation.
I will fix that...

nzakas commented on docs/developer-guide/working-with-rules.md:
```diff
@@ -311,6 +311,121 @@ Keep in mind that comments are technically not a part of the AST and are only at
+ESLint is analyzing code paths while traversing AST.
+You can access that code path objects with five events related to code paths.
+
+#### Details The Code Path Object
```
I'd call this "Code Path Object Details"

nzakas commented on docs/developer-guide/working-with-rules.md:
```diff
@@ -311,6 +311,121 @@ Keep in mind that comments are technically not a part of the AST and are only at
+        },
+
+        /**
+         * This is called when a code path segment was looped.
```
Do you mean using `for` loops, for example?

nzakas commented:
Can you break up this pull request? My browser is struggling to review the code right now.

Also, any update on improving the performance?

mysticatea commented:
- I fixed LogicalExpression's problem, and added more tests for LogicalExpressions.
- I separated commits test fixtures (126 files) and programs (13 files).
- I separated PRs this feature and rules updates.

I will do next:

- Send PRs of rules.
- Add more docs. May I use images? If yes, how do I use it?
- Performance improvements.

ilyavolodin commented:
@mysticatea I don't think breaking up commit into smaller parts worked. Each of those separate commits have 144 modified files.
I'm also not sure how we can use images in documentation. We would need to modify our makefile to copy images to the website repository.

mysticatea commented:
@ilyavolodin Should I separate all tests from this PR? 126 files are test fixtures.

mysticatea commented:
I profiled CPU usage.
But I didn't find special slow function.
Slow functions in code path analysis are small but a great many call, and there are not callers inside of loops.
So I guessed I need to find something that can cache data to local, bit-by-bit...

![image](https://cloud.githubusercontent.com/assets/1937871/10712946/98ec9de6-7ae4-11e5-87e4-ceee669f60bd.png)
![image](https://cloud.githubusercontent.com/assets/1937871/10712943/835e0b7c-7ae4-11e5-9fdd-fb60d7c084e8.png)

mysticatea commented:
@ilyavolodin Ah, sorry, each rule's commit is forked from fd898b1. So it contains files of this PR.
But, those rules don't work without CodePath API.
Umm...

ilyavolodin commented:
I think it's fine if they do not work for now. It would at least give us an ability to review the code. And once the main CodePath API will get merged, you can rebase them.

nzakas commented:
I'm pretty sure images will be copied over for the documentation just fine.

nzakas commented:
@mysticatea you can either figure out a way to speed up code path analysis, or figure out a way to speed up other parts of ESLint to make up for the difference. Both work. :)

mysticatea commented:
I added more documents.
It's big, so I made an independent page: [code-path-analysis.md](https://github.com/mysticatea/eslint/blob/4b82a5ea83374c935c36f1c87a999ad0c20b0978/docs/developer-guide/code-path-analysis.md)

I will try to improvement performance in this weekend.

mysticatea commented:
I modified some functions that a profiler says it's slow.

- `RuleContext` constructor (400ms -> 1ms)
- `getNodeIndent` function (160ms -> 8ms)
- etc...

Then I modified `npm run perf` (`eslint --no-ignore ./tests/performance/jshint.js` --> `eslint --no-ignore lib`) and ran.

As results:

**Original**

```
CPU Speed is 3392 with multiplier 7500000
Performance Run #1:  3337.679209ms
Performance Run #2:  3377.260864ms
Performance Run #3:  3337.500818ms
Performance Run #4:  3295.8551429999998ms
Performance Run #5:  3313.856627ms
Performance budget exceeded: 3337.500818ms (limit: 2211.0849056603774ms)
```

**Added Code Path Analysis**

```
CPU Speed is 3392 with multiplier 7500000
Performance Run #1:  3832.92408ms
Performance Run #2:  3800.501909ms
Performance Run #3:  3816.432429ms
Performance Run #4:  3836.903315ms
Performance Run #5:  3832.869748ms
Performance budget exceeded: 3832.869748ms (limit: 2211.0849056603774ms)
```

500ms slowly... I'm sad....

**Added Code Path Analysis + Improved Performance**

```
CPU Speed is 3392 with multiplier 7500000
Performance Run #1:  3238.026604ms
Performance Run #2:  3136.37427ms
Performance Run #3:  3163.708829ms
Performance Run #4:  3153.404716ms
Performance Run #5:  3131.907554ms
Performance budget exceeded: 3153.404716ms (limit: 2211.0849056603774ms)
```

About 150ms faster then original.

-----

-----

On the other hand, performance for one file linting is still slower.

mysticatea commented:
Single file case (`npm run perf`, target is `./tests/performance/jshint.js`):

**Original**

```
CPU Speed is 3392 with multiplier 7500000
Performance Run #1:  2001.566882ms
Performance Run #2:  2000.67372ms
Performance Run #3:  2004.677403ms
Performance Run #4:  1999.6082040000001ms
Performance Run #5:  1998.1629659999999ms
Performance budget ok:  2000.67372ms (limit: 2211.0849056603774ms)
```

**Added Code Path Analysis**

```
CPU Speed is 3392 with multiplier 7500000
Performance Run #1:  2216.400856ms
Performance Run #2:  2250.396248ms
Performance Run #3:  2223.547359ms
Performance Run #4:  2219.674072ms
Performance Run #5:  2228.813362ms
Performance budget exceeded: 2223.547359ms (limit: 2211.0849056603774ms)
```

**Added Code Path Analysis + Improved Performance**

```
CPU Speed is 3392 with multiplier 7500000
Performance Run #1:  2189.549854ms
Performance Run #2:  2186.523547ms
Performance Run #3:  2199.52586ms
Performance Run #4:  2181.768569ms
Performance Run #5:  2175.863559ms
Performance budget ok:  2186.523547ms (limit: 2211.0849056603774ms)
```

It's hard to improve more...

ilyavolodin commented:
Nice! That's a pretty good performance optimization, especially on multiple files.

ilyavolodin commented on docs/developer-guide/code-path-analysis.md:
```diff
@@ -0,0 +1,554 @@
+
+* `id` (`string`) - An unique string. Respective rules can use `id` to save additional information for each segment.
+* `nextSegments` (`CodePathSegment[]`) - The next segments. If forking, there are two or more. If final, there is nothing.
+* `prevSegments` (`CodePathSegment[]`) - The previous segments. If merging, there are two or more. If final, there is nothing.
```
I assume "If final, there is nothing." was copied from above? I assume this was supposed to say "If entry, there is nothing" right?

mysticatea replied:
Yeah, I mistook!

IanVS commented:
Hmm, I wish commit comments showed up better here.  Sorry for that.  I can't view the whole diff because it's too large. :open_mouth:

nzakas commented:
Wow, nice! We can probably live with the hit on a single file so long as linting multiple files isn't affected. Nice work!

nzakas commented:
@mysticatea can you rebase? I think we are ready to merge.

mysticatea commented:
I'm glad!
I squashed and rebased.

lo1tuma commented:
Do we want the performance tweaks to be part of this PR? The related changes to `RuleContext` is changing the behavior of it (e.g. the `ecmaFeatures` object isn’t frozen anymore).

mysticatea commented:
OK, I will separate the performance tweaks from this PR.

nzakas commented:
Okay, the perf piece is merged. Is there anything else we can split out from this PR? I'm having a hard time loading the diffs because there are so many files. :(

mysticatea commented:
I can split out test fixtures (126 files) and documents (26 files).

nzakas commented:
Hmm ok, we'll just merge this all together, then. We will wait until Monday just to be sure we don't need a patch release.
